<a href="https://colab.research.google.com/github/nagaharshitha20/Credit_Card_Fraud_Detection/blob/main/Credit_Card_Fraud_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np

file_path = "/content/drive/MyDrive/Colab Notebooks/creditcard.csv"
df = pd.read_csv(file_path)

df.head()


Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
print("Shape of the dataset:",df.shape)

Shape of the dataset: (284807, 31)


In [ ]:
print("Column names:")
print(df.columns.tolist())

Column names:
['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount', 'Class']


In [ ]:
print("Dataset info")
print(df.info())

Dataset info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 2

In [ ]:
print("\nMissing values in each column:")
print(df.isna().sum())


Missing values in each column:
Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64


In [ ]:
print("\nClass distribution (0 = legit, 1 = fraud):")
print(df['Class'].value_counts())


Class distribution (0 = legit, 1 = fraud):
Class
0    284315
1       492
Name: count, dtype: int64


In [ ]:
print("\nClass distribution in percentage:")
print(df['Class'].value_counts(normalize=True) * 100)


Class distribution in percentage:
Class
0    99.827251
1     0.172749
Name: proportion, dtype: float64


In [ ]:
# Separate features (X) and target (y)

X = df.drop("Class", axis=1)   # all columns except Class
y = df["Class"]                # only the Class column

X.head(), y.head()


(   Time        V1        V2        V3        V4        V5        V6        V7  \
 0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
 1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
 2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
 3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
 4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
 
          V8        V9  ...       V20       V21       V22       V23       V24  \
 0  0.098698  0.363787  ...  0.251412 -0.018307  0.277838 -0.110474  0.066928   
 1  0.085102 -0.255425  ... -0.069083 -0.225775 -0.638672  0.101288 -0.339846   
 2  0.247676 -1.514654  ...  0.524980  0.247998  0.771679  0.909412 -0.689281   
 3  0.377436 -1.387024  ... -0.208038 -0.108300  0.005274 -0.190321 -1.175575   
 4 -0.270533  0.817739  ...  0.408542 -0.009431  0.798278 -0.137458  0.141267   
 
         V25      

In [ ]:
from sklearn.model_selection import train_test_split

# Stratified split to maintain fraud ratio in both sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

print("Train shape:", X_train.shape, y_train.shape)
print("Test shape:", X_test.shape, y_test.shape)

print("\nFraud ratio in train:")
print(y_train.value_counts(normalize=True) * 100)

print("\nFraud ratio in test:")
print(y_test.value_counts(normalize=True) * 100)


Train shape: (227845, 30) (227845,)
Test shape: (56962, 30) (56962,)

Fraud ratio in train:
Class
0    99.827075
1     0.172925
Name: proportion, dtype: float64

Fraud ratio in test:
Class
0    99.827955
1     0.172045
Name: proportion, dtype: float64


In [ ]:
from sklearn.preprocessing import StandardScaler

# Create the scaler object
scaler = StandardScaler()

# Fit ONLY on training data
scaler.fit(X_train)

# Transform both train and test datasets
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Checking shapes to confirm everything is good
print("Scaled X_train shape:", X_train_scaled.shape)
print("Scaled X_test shape:", X_test_scaled.shape)


Scaled X_train shape: (227845, 30)
Scaled X_test shape: (56962, 30)


# -------------------------------
# SMALL BALANCED DATASET
# -------------------------------


In [ ]:
# Create small balanced dataset: 492 fraud + 492 non-fraud

# Separate fraud and non-fraud rows
fraud = df[df['Class'] == 1]
non_fraud = df[df['Class'] == 0].sample(n=len(fraud), random_state=42)

# Combine them
balanced_df = pd.concat([fraud, non_fraud], axis=0)

# Shuffle the rows
balanced_df = balanced_df.sample(frac=1, random_state=42).reset_index(drop=True)

# Show shapes
print("Balanced dataset shape:", balanced_df.shape)

# Check class distribution
print("\nClass distribution:")
print(balanced_df['Class'].value_counts())

print("\nClass distribution in percentage:")
print(balanced_df['Class'].value_counts(normalize=True) * 100)


Balanced dataset shape: (984, 31)

Class distribution:
Class
0    492
1    492
Name: count, dtype: int64

Class distribution in percentage:
Class
0    50.0
1    50.0
Name: proportion, dtype: float64


Here we need not to use stratification as we did for full data set because this data set is already balanced with equal fraud and non fraud but still we can keep stratify=y for safety



In [ ]:
from sklearn.model_selection import train_test_split

# Separate X and y for balanced dataset
X_balanced = balanced_df.drop("Class", axis=1)
y_balanced = balanced_df["Class"]

# Split into train and test
X_train_small, X_test_small, y_train_small, y_test_small = train_test_split(
    X_balanced, y_balanced,
    test_size=0.2,
    stratify=y_balanced,      # optional but safe
    random_state=42
)

print("Train shape:", X_train_small.shape, y_train_small.shape)
print("Test shape:", X_test_small.shape, y_test_small.shape)

print("\nClass ratio in TRAIN:")
print(y_train_small.value_counts(normalize=True) * 100)

print("\nClass ratio in TEST:")
print(y_test_small.value_counts(normalize=True) * 100)


Train shape: (787, 30) (787,)
Test shape: (197, 30) (197,)

Class ratio in TRAIN:
Class
1    50.063532
0    49.936468
Name: proportion, dtype: float64

Class ratio in TEST:
Class
0    50.253807
1    49.746193
Name: proportion, dtype: float64


this is a new dataset with 984 rows, not the full dataset.
Scaling must be done on the dataset you actually use for training.So we must:

Create a new scaler

Fit it only on training data

Transform both training and testing data

In [ ]:
from sklearn.preprocessing import StandardScaler

# Create scaler
scaler_small = StandardScaler()

# Fit only on training data
scaler_small.fit(X_train_small)

# Transform training and testing data
X_train_small_scaled = scaler_small.transform(X_train_small)
X_test_small_scaled = scaler_small.transform(X_test_small)

# Check shapes
print("Scaled X_train_small shape:", X_train_small_scaled.shape)
print("Scaled X_test_small shape:", X_test_small_scaled.shape)


Scaled X_train_small shape: (787, 30)
Scaled X_test_small shape: (197, 30)


Model building

1.Logistic Reg as baseline model


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Create Logistic Regression model
log_reg = LogisticRegression(max_iter=1000, class_weight='balanced')

# Train (fit) on the small scaled training data
log_reg.fit(X_train_small_scaled, y_train_small)

# Predict on the small scaled test data
y_pred_lr = log_reg.predict(X_test_small_scaled)

# Evaluation metrics
print("Accuracy:", accuracy_score(y_test_small, y_pred_lr))

print("\nClassification Report:")
print(classification_report(y_test_small, y_pred_lr))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test_small, y_pred_lr))


Accuracy: 0.934010152284264

Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.95      0.94        99
           1       0.95      0.92      0.93        98

    accuracy                           0.93       197
   macro avg       0.93      0.93      0.93       197
weighted avg       0.93      0.93      0.93       197


Confusion Matrix:
[[94  5]
 [ 8 90]]


In [ ]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Create the SVM model
svm_model = SVC(kernel='rbf', class_weight='balanced')

# Train the model
svm_model.fit(X_train_small_scaled, y_train_small)

# Predict on the test set
y_pred_svm = svm_model.predict(X_test_small_scaled)

# Evaluation
print("Accuracy:", accuracy_score(y_test_small, y_pred_svm))

print("\nClassification Report:")
print(classification_report(y_test_small, y_pred_svm))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test_small, y_pred_svm))


Accuracy: 0.9238578680203046

Classification Report:
              precision    recall  f1-score   support

           0       0.89      0.97      0.93        99
           1       0.97      0.88      0.92        98

    accuracy                           0.92       197
   macro avg       0.93      0.92      0.92       197
weighted avg       0.93      0.92      0.92       197


Confusion Matrix:
[[96  3]
 [12 86]]


Logistic reg>SVM


RANDOM FOREST (Best model)


In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Create NEW Random Forest model (fresh variable)
rf_final = RandomForestClassifier(
    n_estimators=500,
    max_depth=None,
    class_weight='balanced',
    random_state=42
)

# Fit on small scaled data
rf_final.fit(X_train_small_scaled, y_train_small)

# Predict on scaled test data
rf_pred_final = rf_final.predict(X_test_small_scaled)

# Evaluation
print("Accuracy:", accuracy_score(y_test_small, rf_pred_final))

print("\nClassification Report:")
print(classification_report(y_test_small, rf_pred_final))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test_small, rf_pred_final))


Accuracy: 0.9390862944162437

Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.96      0.94        99
           1       0.96      0.92      0.94        98

    accuracy                           0.94       197
   macro avg       0.94      0.94      0.94       197
weighted avg       0.94      0.94      0.94       197


Confusion Matrix:
[[95  4]
 [ 8 90]]


In [ ]:
print(rf_model)


RandomForestClassifier(class_weight='balanced', n_estimators=300,
                       random_state=42)
